olp-cpp-sdk-authentication/src/AuthenticationClient.cpp

```diff
@@ -450,206 +451,124 @@ AuthenticationClient::Impl::~Impl() { pending_requests_->CancelAllAndWait(); }
+         GenerateHeader(credentials, url, timestamp)}};
+
+    auto auth_response = client.CallApi(
+        kOauthEndpoint, "POST", {}, std::move(headers), {},
+        GenerateClientBody(properties), kApplicationJson, context);
 
-        return;
+    const auto status = auth_response.status;
+
+    if (status < 0) {
+      auto error_message = auth_response.response.str();
+
+      if (status == static_cast<int>(olp::http::ErrorCode::CANCELLED_ERROR)) {
+        return {{status, error_message}};
       }
 
-      auto document = std::make_shared<rapidjson::Document>();
-      rapidjson::IStreamWrapper stream(*payload);
-      document->ParseStream(stream);
-
-      std::shared_ptr<SignInResultImpl> resp_impl =
-          std::make_shared<SignInResultImpl>(response_status, error_msg,
-                                             document);
-      SignInResult response(resp_impl);
-
-      if (response_status == http::HttpStatusCode::OK) {
-        // Cache the response
-        cache->locked([credentials, &response](
-                          utils::LruCache<std::string, SignInResult>& c) {
-          return c.InsertOrAssign(credentials.GetKey(), response);
-        });
+      auto cached_response = client_token_cache_->locked(
+          [credentials](utils::LruCache<std::string, SignInResult>& c)
+              -> boost::optional<SignInResult> {
+            auto it = c.Find(credentials.GetKey());
+            return it != c.end() ? boost::make_optional(it->value())
+                                 : boost::none;
+          });
+
+      if (!cached_response) {
+        return {{status, error_message}};
       }
-      callback(response);
-    };
-
-    context.ExecuteOrCancelled(
-        [&]() {
-          auto send_outcome = settings_.network_request_handler->Send(
-              request, payload, network_callback);
-          if (!send_outcome.IsSuccessful()) {
-            std::string error_message =
-                ErrorCodeToString(send_outcome.GetErrorCode());
-            callback(AuthenticationError(
-                static_cast<int>(send_outcome.GetErrorCode()), error_message));
-            return client::CancellationToken();
-          }
-          auto request_id = send_outcome.GetRequestId();
-          return client::CancellationToken([weak_network, request_id]() {
-            auto network = weak_network.lock();
 
-            if (network) {
-              network->Cancel(request_id);
-            }
+      return *cached_response;
+    }
+
+    auto document = std::make_shared<rapidjson::Document>();
+    rapidjson::IStreamWrapper stream(auth_response.response);
+    document->ParseStream(stream);
+
+    std::shared_ptr<SignInResultImpl> resp_impl =
+        std::make_shared<SignInResultImpl>(status, "", document);
+    SignInResult response(resp_impl);
+
+    if (status == http::HttpStatusCode::OK) {
+      // Cache the response
+      client_token_cache_->locked(
+          [credentials,
```

Comment:
```suggestion
          [&
```
No need to copy on a sync call.